sw/device/silicon_creator/manuf/base/ft_personalize.c

```diff
@@ -861,17 +864,25 @@ static status_t write_cert_to_flash_info_page(
     return OUT_OF_RANGE();
   }
 
-  memset(cert_buffer, 0, cert_write_size_bytes);
-
   // Copy the actual certificate data into the cert buffer.
-  // This is necessary because flash_ctrl_info_write() requires the
-  // data source pointer to be word-aligned. The input cert_data
-  // pointer might not meet this alignment requirement, whereas
-  // cert_buffer is expected to be world-aligned.
-  memcpy(cert_buffer, cert_data, block->obj_size);
+  memcpy(cert_buffer + page_offset, cert_data, block->obj_size);
+
+  return OK_STATUS();
+}
+
+static status_t write_digest_to_cert_buffer(
+    const cert_flash_info_layout_t *layout, uint32_t page_offset) {
+  base_printf("Digesting %s page ...\n", layout->group_name);
+  const size_t digest_offset =
+      FLASH_CTRL_PARAM_BYTES_PER_PAGE - sizeof(hmac_digest_t);
+
+  if (page_offset > digest_offset) {
+    LOG_ERROR("%s digest did not fit into the info page.", layout->group_name);
+    return OUT_OF_RANGE();
+  }
 
-  TRY(flash_ctrl_info_write(layout->info_page, page_offset,
-                            cert_write_size_words, cert_buffer));
+  hmac_sha256(cert_buffer, digest_offset,
+              (hmac_digest_t *)(&cert_buffer[digest_offset]));
```

Comment:
Similarly, if we use a struct to represent the cert page, we can eliminate the explicit offset calculation and checking here in favor of the intrinsic offsets & sizeof's known by the compiler.